node/impl/full/eth.go

```diff
@@ -1039,6 +1041,163 @@ func (a *EthModule) EthTraceTransaction(ctx context.Context, txHash string) ([]*
+	head := a.Chain.GetHeaviestTipSet()
+
+	blockValue := "latest"
+	if block != nil {
+		blockValue = *block
+	}
+
+	switch blockValue {
+	case "earliest":
+		return 0, xerrors.Errorf("block param \"earliest\" is not supported")
+	case "pending":
+		return ethtypes.EthUint64(head.Height()), nil
+	case "latest":
+		parent, err := a.Chain.GetTipSetFromKey(ctx, head.Parents())
+		if err != nil {
+			return 0, fmt.Errorf("cannot get parent tipset")
+		}
+		return ethtypes.EthUint64(parent.Height()), nil
+	case "safe":
+		latestHeight := head.Height() - 1
+		safeHeight := latestHeight - ethtypes.SafeEpochDelay
+		return ethtypes.EthUint64(safeHeight), nil
+	default:
+		blockNum, err := ethtypes.EthUint64FromHex(blockValue)
+		if err != nil {
+			return 0, xerrors.Errorf("cannot parse fromBlock: %w", err)
+		}
+		return blockNum, err
+	}
+}
+
+func (a *EthModule) EthTraceFilter(ctx context.Context, filter ethtypes.EthTraceFilterCriteria) ([]*ethtypes.EthTraceFilterResult, error) {
+	fromBlock, err := a.EthBlockNumberFromString(ctx, filter.FromBlock)
+	if err != nil {
+		return nil, xerrors.Errorf("cannot parse fromBlock: %w", err)
+	}
+
+	toBlock, err := a.EthBlockNumberFromString(ctx, filter.ToBlock)
+	if err != nil {
+		return nil, xerrors.Errorf("cannot parse toBlock: %w", err)
+	}
+
+	var results []*ethtypes.EthTraceFilterResult
+
+	if filter.Count != nil {
+		// If filter.Count is specified and it is 0, return an empty result set immediately.
+		if *filter.Count == 0 {
+			return []*ethtypes.EthTraceFilterResult{}, nil
+		}
+
+		// If filter.Count is specified and is greater than the EthTraceFilterMaxResults config return error
+		if *filter.Count > a.EthTraceFilterMaxResults {
+			return nil, xerrors.Errorf("invalid response count")
+		}
+	}
+
+	traceCounter := ethtypes.EthUint64(0)
+	for blkNum := fromBlock; blkNum <= toBlock; blkNum++ {
+		blockTraces, err := a.EthTraceBlock(ctx, strconv.FormatUint(uint64(blkNum), 10))
+		if err != nil {
+			return nil, xerrors.Errorf("cannot get trace for block %d: %w", blkNum, err)
+		}
+
+		for _, blockTrace := range blockTraces {
+			match, err := matchFilterCriteria(blockTrace, filter, filter.FromAddress, filter.ToAddress)
+			if err != nil {
+				return nil, xerrors.Errorf("cannot match filter for block %d: %w", blkNum, err)
+			}
+			if match {
+				traceCounter++
+				if filter.After != nil && traceCounter <= *filter.After {
+					continue
+				}
+
+				txTrace := ethtypes.EthTraceFilterResult{
+					EthTrace:            blockTrace.EthTrace,
+					BlockHash:           blockTrace.BlockHash,
+					BlockNumber:         blockTrace.BlockNumber,
+					TransactionHash:     blockTrace.TransactionHash,
+					TransactionPosition: blockTrace.TransactionPosition,
+				}
+				results = append(results, &txTrace)
+
+				// If Count is specified, limit the results
+				if filter.Count != nil && ethtypes.EthUint64(len(results)) >= *filter.Count {
+					return results, nil
+				}
```

Comment:
```suggestion
				if filter.Count != nil && ethtypes.EthUint64(len(results)) >= *filter.Count {
					return results, nil
				} else if filter.Count == nil && len(results) > a.EthTraceFilterMaxResults {
					return nil, xerrors.Errorf("too many results, maximum supported is %d, try paginating requests with After and Count", *filter.Count, a.EthTraceFilterMaxResults)
				}
```

something like this? to account for the case where you don't specify a count but you end up matching way too many responses.

Reply:
Yeah this is a bug. Need to account for a nil count.